core/src/main/scala/org/apache/spark/memory/StaticMemoryManager.scala

```diff
@@ -62,17 +64,34 @@ private[spark] class StaticMemoryManager(
     }
   }
 
-  override def acquireUnrollMemory(blockId: BlockId, numBytes: Long): Boolean = synchronized {
-    val currentUnrollMemory = storageMemoryPool.memoryStore.currentUnrollMemory
-    val freeMemory = storageMemoryPool.memoryFree
+  override def acquireUnrollMemory(blockId: BlockId, numBytes: Long): Boolean = {
     // When unrolling, we will use all of the existing free memory, and, if necessary,
     // some extra space freed from evicting cached blocks. We must place a cap on the
     // amount of memory to be evicted by unrolling, however, otherwise unrolling one
     // big block can blow away the entire cache.
-    val maxNumBytesToFree = math.max(0, maxUnrollMemory - currentUnrollMemory - freeMemory)
-    // Keep it within the range 0 <= X <= maxNumBytesToFree
-    val numBytesToFree = math.max(0, math.min(maxNumBytesToFree, numBytes - freeMemory))
-    storageMemoryPool.acquireMemory(blockId, numBytes, numBytesToFree)
+    val (freeMemoryAcquired, numBytesToFree) = synchronized {
+      val currentUnrollMemory = storageMemoryPool.memoryStore.currentUnrollMemory
+      val freeMemory = storageMemoryPool.memoryFree
+      val freeMemoryAcquired = math.min(numBytes, freeMemory)
+      val res = storageMemoryPool.acquireMemory(blockId, freeMemoryAcquired, numBytesToFree = 0)
+      assert(res, "failed to acquire free storage memory")
+      val maxNumBytesToFree = math.max(0, maxUnrollMemory - currentUnrollMemory - freeMemory)
+      // Keep it within the range 0 <= X <= maxNumBytesToFree
+      val numBytesToFree = math.min(numBytes - freeMemoryAcquired, maxNumBytesToFree)
+      (freeMemoryAcquired, numBytesToFree)
+    }
+    try {
+      if (storageMemoryPool.acquireMemory(blockId, numBytes - freeMemoryAcquired, numBytesToFree)) {
+        true
+      } else {
+        storageMemoryPool.releaseMemory(freeMemoryAcquired)
+        false
+      }
+    } catch {
+      case NonFatal(e) =>
+        storageMemoryPool.releaseMemory(freeMemoryAcquired)
```

Comment:
it's kind of weird that it's possible to call `releaseMemory` twice here. If the first one fails at the end (after actually releasing memory) it will come here and possibly release the memory again (and then failing again). A better way to write this would be something like:
```
// Acquire the rest of the memory outside a synchronized block.
// This call may evict existing blocks if necessary.
val acquired =
  try {
    storageMemoryPool.acquire(...)
  } catch {
    case NonFatal(e) =>
      storageMemoryPool.releaseMemory(...)
      throw e
  }
if (!acquired) {
  storageMemoryPool.releaseMemory(...)
}
acquired
```